# Spam Detector

In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [2]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

*While Logistic Regression is computationally faster and requires less training, Random Forest in my opinion is going to perform better than a Logistic Regression as a random forest is going to involve multiple decision trees / paths to reach a prediction and can provide importance feature analysis, helping us identify the feature(s) most likely to influence the target.*

## Split the Data into Training and Testing Sets

In [13]:
# Create the labels set `y` and features DataFrame `X`

# Creating a features dataframe
X = data.copy()
X.drop("spam", axis=1, inplace=True)

# Verifyings that the dataframe is without the target column 'spam'
display(X.head())

# Create the 'y' target data set
y = data['spam']

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
0             0.00            0.00  ...                   0.0         0.00   
1             0.00            0.94  ...                   0.0         0.00   
2             0.64            0.25  ...                   0.0         0.01   
3             0.31            0.63  ...                   0.0         0.00   
4             0.31            0.63  ...                   0.0         0.00   

   char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0        0.000          0.0        0.778        0.000        0.000   
1        0.132          0.0        0.372        0.180        0.048   
2        0.143          0.0        0.276        0.184        0.010   
3        0.137          0.0        0.137        0.000        0.000   
4        0.135          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  
0                       278  
1                      1028  
2                      2259  
3                       191  
4                       191  

[5 rows x 57 columns]

In [14]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [28]:
# Defining the 'y' target vector
y = data['spam'].to_numpy()
y[:5]

array([1, 1, 1, 1, 1])

In [16]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [17]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance and fit the model 
X_scaler = StandardScaler().fit(X_train)


In [18]:
# Scale the training data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [21]:
# Create and fit a Logistic Regression model
from sklearn.linear_model import LogisticRegression

logistic_regression_model = LogisticRegression(random_state=1, max_iter=120)

lr_model = logistic_regression_model.fit(X_train_scaled, y_train)

In [30]:
# Make and save testing predictions with the trained Logistic Regression model using the test data
lr_predictions = lr_model.predict(X_test_scaled)

# Review the predictions
lr_predictions

array([0, 0, 0, ..., 1, 0, 0])

In [31]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
lr_acc_score = accuracy_score(y_test,lr_predictions)

lr_acc_score

0.9226759339704604

## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. Use a `random_state` of `1`. 

In [32]:
# Create and fit a Random Forest Classifier model
from sklearn.ensemble import RandomForestClassifier

rfcl_model = RandomForestClassifier(random_state=1, n_estimators=120)

In [34]:
# Make and save testing predictions with the trained Random Forest Classifier model using the test data
rfcl_model = rfcl_model.fit(X_train_scaled,y_train)

# Review the predictions
rfcl_predictions = rfcl_model.predict(X_test_scaled)

rfcl_predictions

array([0, 0, 0, ..., 1, 0, 0])

In [35]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
rfcl_acc_score = accuracy_score(y_test, rfcl_predictions)

rfcl_acc_score

0.9609035621198957

In [38]:
# Feature Importance
rfcl_features = rfcl_model.feature_importances_

rfcl_features_importance_sorted = sorted(zip(rfcl_model.feature_importances_, X.columns), reverse=True)
rfcl_features_importance_sorted[:10]

[(0.12238196779660651, 'char_freq_$'),
 (0.099207797736144, 'char_freq_!'),
 (0.08038903871306023, 'word_freq_remove'),
 (0.06957126895480324, 'capital_run_length_average'),
 (0.06644647904791073, 'word_freq_free'),
 (0.05765665333975998, 'capital_run_length_longest'),
 (0.048637651806908835, 'word_freq_your'),
 (0.043790620721814874, 'word_freq_hp'),
 (0.037352214649392444, 'capital_run_length_total'),
 (0.03243719411340246, 'word_freq_you')]

## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.

*Random Forest performed better as predicted (LR accuracy | 0.92 vs. RFCL | 0.96). While the Random Forest Classifier Model took a little longer to predict and proces (LR |0.0s vs. RCFL 0.6s), it still proved to be a little bit more accurate than Logistic Regression Model*